### PR TITLE
#13: QueryResultRow conversion for Product types

### LINKED ISSUE #13: QueryResultRow conversion for Product types
## Background
At the moment it's difficult to create test assertions of the db results given the fact that one needs to extract fields from the QueryResultRow object. I propose to enhance capabilities of the library to allow for automatic instance creation for product types.

## Feature
Balta will be able to create instances of product types from QueryResultRow

## Example 
```scala
class GetActorByIdTest extends DBTestSuite {

  test("Get actor by id - actor with given id exists EXISTING") {
    val expectedActor = Actor(actorId = 1, firstName = "Malin", lastName = "Akerman")

    function("runs.get_actor_by_id")
      .setParam("i_actor_id", 1)
      .execute { queryResult =>
        assert(queryResult.hasNext)
        val row = queryResult.next()
        assert(row.getInt("actor_id").get == expectedActor.actorId)
        assert(row.getString("first_name").get == expectedActor.firstName)
        assert(row.getString("last_name").get == expectedActor.lastName)
        assert(!queryResult.hasNext)
      }
  }

  import za.co.absa.balta.classes.DefaultColumnNameMappers.SnakeCaseForCamelCaseMapper

  test("Get actor by id - actor with given id exists PROPOSAL") {
    val expectedActor = Actor(actorId = 1, firstName = "Malin", lastName = "Akerman")

    function("runs.get_actor_by_id")
      .setParam("i_actor_id", 1)
      .execute { queryResult =>
        assert(queryResult.hasNext)
        val row = queryResult.next()
        val actualActor = row.toProductType[Actor]
        assert(expectedActor == actualActor)
        assert(!queryResult.hasNext)
      }
  }

}
```

## Proposed Solution for Discussion
An example of possible solution (needs to be tested for various scenarios, only tested on the test example above)

```scala
package za.co.absa.balta.classes

import org.postgresql.util.PGobject

import java.sql.{Date, ResultSet, Time}
import java.time.{Instant, OffsetDateTime}
import java.util.UUID

import scala.reflect.runtime.universe._
import scala.reflect.runtime.currentMirror

class QueryResultRow private[classes](val resultSet: ResultSet) extends AnyVal {

  // existing code not included

  def toProductType[T <: Product : TypeTag](implicit mapper: ColumnNameMapper): T = {
    val tpe = typeOf[T]
    val classSymbol = tpe.typeSymbol.asClass
    val classMirror = currentMirror.reflectClass(classSymbol)
    val constructorSymbol = tpe.decl(termNames.CONSTRUCTOR)
    val defaultConstructor =
      if (constructorSymbol.isMethod) constructorSymbol.asMethod
      else {
        val ctors = constructorSymbol.asTerm.alternatives
        ctors.map(_.asMethod).find(_.isPrimaryConstructor).get
      }
    val constructorMirror = classMirror.reflectConstructor(defaultConstructor)
    val params = defaultConstructor.paramLists.flatten.map { param =>
      val name = param.name.decodedName.toString
      val columnName = mapper.map(name)
      val value = resultSet.getObject(columnName)
      if (resultSet.wasNull()) null else value
    }
    constructorMirror(params: _*).asInstanceOf[T]
  }

}

trait ColumnNameMapper {
  def map(name: String): String
}

object DefaultColumnNameMappers {

  // when database columns use the same names as product type
  implicit object IdentityMapper extends ColumnNameMapper {
    def map(name: String): String = name
  }

  // when database columns use upper case and product type uses lower case
  implicit object UpperCaseForLowerCaseMapper extends ColumnNameMapper {
    def map(name: String): String = name.toUpperCase
  }

  // when database columns use lower case and product type uses upper case
  implicit object LowerCaseForUpperCaseMapper extends ColumnNameMapper {
    def map(name: String): String = name.toLowerCase
  }

  // when database columns use snake_case and product type uses camelCase
  implicit object SnakeCaseForCamelCaseMapper extends ColumnNameMapper {
    def map(name: String): String = {
      val regex = "([a-z])([A-Z]+)".r
      val replacement = "$1_$2"
      regex.replaceAllIn(name, replacement).toLowerCase
    }
  }

  // when database columns use kebab-case and product type uses camelCase
  implicit object KebabCaseForCamelCaseMapper extends ColumnNameMapper {
    def map(name: String): String = {
      val regex = "([a-z])([A-Z]+)".r
      val replacement = "$1-$2"
      regex.replaceAllIn(name, replacement).toLowerCase
    }
  }

}
```

### PR DESCRIPTION
**Pull request by benedeki**
* implicit class that adds the ability to convert `QueryResultRow` to a product type
* implicit classes that add `getOrThrow` methods to `Option` and `Map`
* copied `NamingConvention` classes from Fa-DB
* added `sbt testAll` alias
* enhanced the README.md to include some basic classes of the library used for DB testing

Closes #13

### CODE REVIEW

**Review by miroslavpojer — COMMENTED** [review]
- [ ] pull
- [ ] code review
- [ ] run all test types

File: balta/src/test/scala/za/co/absa/db/mag/naming/implementations/AsIsNamingUnitTests.scala [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+package za.co.absa.db.mag.naming.implementations
+
+import org.scalatest.funsuite.AnyFunSuiteLike
+import org.scalatest.matchers.should.Matchers
+import za.co.absa.db.mag.naming.LettersCase
+
+class AsIsNamingUnitTests extends AnyFunSuiteLike with Matchers {
+
+  val asIsNaming = new AsIsNaming(LettersCase.AsIs)
```
[the remark this answers is not in the document]
Comment (benedeki) [reply balta/src/test/scala/za/co/absa/db/mag/naming/implementations/AsIsNamingUnitTests.scala:25]: Not sure I understand. I am creating an instance of a class here, not a new type.

File: balta/src/test/scala/za/co/absa/db/mag/naming/implementations/SnakeCaseNamingUnitTests.scala [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+package za.co.absa.db.mag.naming.implementations
+
+import org.scalatest.matchers.should.Matchers
+import org.scalatest.wordspec.AnyWordSpec
+import za.co.absa.db.mag.naming.LettersCase.{AsIs, LowerCase, UpperCase}
+
+class SnakeCaseNamingUnitTests extends AnyWordSpec with Matchers {
+  private class ThisIsATestClass
+  private val testInstance = new ThisIsATestClass()
```
[the remark this answers is not in the document]
Comment (benedeki) [reply balta/src/test/scala/za/co/absa/db/mag/naming/implementations/SnakeCaseNamingUnitTests.scala:25]: Whet exactly you want to expand there? The suggestion is already present, isn't it? I don't see a difference from the current tests, sorry.

File: balta/src/test/scala/za/co/absa/db/mag/naming/implementations/AsIsNamingUnitTests.scala (hunk 1 above)
[the remark this answers is not in the document]
Comment (miroslavpojer) [reply balta/src/test/scala/za/co/absa/db/mag/naming/implementations/AsIsNamingUnitTests.scala:25]: New AsIsNaming and then define LetterCase.AsIs in constructor - it looks as saying same thing twice.
At least, I would expect that LetterCase is the default value of AsIsNaming.

File: balta/src/test/scala/za/co/absa/db/mag/naming/implementations/SnakeCaseNamingUnitTests.scala (hunk 2 above)
[the remark this answers is not in the document]
Comment (miroslavpojer) [reply balta/src/test/scala/za/co/absa/db/mag/naming/implementations/SnakeCaseNamingUnitTests.scala:25]: Now, it is solved in the code. Not sure what was source of confusion.

File: balta/src/test/scala/za/co/absa/db/mag/naming/implementations/AsIsNamingUnitTests.scala (hunk 1 above)
[the remark this answers is not in the document]
Comment (benedeki) [reply balta/src/test/scala/za/co/absa/db/mag/naming/implementations/AsIsNamingUnitTests.scala:25]: NamingConvention is how names are converted e.g. TitleCase to SnakeCase. AsIsNaming is the implementation, that doesn't change the input in its structure.
LetterCase is the specification how the string case changes e.g. UPPER CASE. AsIs letter case means no case change.
I actually bellieve AsIs letter case is the default for AsIsBaming. But being explicit in tests I consider better then to relly on defaults.

**Review by miroslavpojer — APPROVED** [review]

File: README.md [hunk 3]
```diff
@@ -27,17 +27,71 @@ It's a natural complement to the use of [Fa-Db library](https://github.com/AbsaO
 
  * The transaction start and rollback are done automatically before or after the execution respectively of the `test` function provided
 
-Advantages of this approach is that the tests repeateble, they are isolated from each other and the database is always 
+Advantages of this approach is that the tests are repeateble, they are isolated from each other and the database is always
```
Comment (lsulak) [inline README.md:30]:
```suggestion
Advantages of this approach is that the tests are repeatable, they are isolated from each other and the database is always 
```

File: README.md [hunk 4]
```diff
@@ -27,17 +27,71 @@ It's a natural complement to the use of [Fa-Db library](https://github.com/AbsaO
...
 
+## How to write tests
+
+### [`DBTestSuite`](https://github.com/AbsaOSS/balta/blob/master/balta/src/main/scala/za/co/absa/db/balta/DBTestSuite.scala) class
+
+The foundation is the [`DBTestSuite`](https://github.com/AbsaOSS/balta/blob/master/balta/src/main/scala/za/co/absa/db/balta/DBTestSuite.scala) 
+class that provides the necessary setup and teardown for the tests to each run in its own transaction. It is an 
+enhancement class to standard ScalaTest `AnyFunSuite` class.
+
+Besides that, it provides easy access to tables, query them and insert data into them.
```
Comment (lsulak) [inline README.md:42]:
```suggestion
Besides that, it provides easy access to tables, queries them and inserts data into them.
```

File: balta/src/main/scala/za/co/absa/db/balta/classes/QueryResultRow.scala [hunk 5]
```diff
@@ -35,7 +36,10 @@ class QueryResultRow private[classes](val rowNumber: Int,
                                       private val columnNames: FieldNames) {
 
   def columnCount: Int = fields.length
-  def columnNumber(columnLabel: String): Int = columnNames(columnLabel.toLowerCase)
+  def columnNumber(columnLabel: String): Int = {
```
Comment (lsulak) [inline balta/src/main/scala/za/co/absa/db/balta/classes/QueryResultRow.scala:39]: why `Label` terminology?

File: balta/src/test/resources/db/postgres/08_testing.simple_function.sql [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+    OUT ignored_data    TEXT
+) RETURNS record AS
+$$
+-------------------------------------------------------------------------------
+--
+-- Function: testing.simple_function(1)
+--      Function returning static data for testing purposes
+--
+-- Parameters:
+--      i_return_data   - flag if data are to be returned or not         -
```
Comment (lsulak) [inline balta/src/test/resources/db/postgres/08_testing.simple_function.sql:32]:
```suggestion
--      i_return_data   - flag if data are to be returned or not
```

File: balta/src/main/scala/za/co/absa/db/balta/classes/QueryResultRow.scala (hunk 5 above)
Comment (benedeki) [reply balta/src/main/scala/za/co/absa/db/balta/classes/QueryResultRow.scala:39]: Original terminology from JDBC

Comment (lsulak) [reply balta/src/main/scala/za/co/absa/db/balta/classes/QueryResultRow.scala:39]: but it's column name right?

Comment (lsulak) [reply balta/src/main/scala/za/co/absa/db/balta/classes/QueryResultRow.scala:39]: Hmm I read here that it's this: https://bugs.mysql.com/bug.php?id=35610

so `alias` or `column name`. Not sure if in our case it's always only column name?

Comment (benedeki) [reply balta/src/main/scala/za/co/absa/db/balta/classes/QueryResultRow.scala:39]: Look at the `ResultSet` class methods.

Comment (lsulak) [reply balta/src/main/scala/za/co/absa/db/balta/classes/QueryResultRow.scala:39]: Hmm I still don't fully know the answer if you think deeper into it :D 

https://docs.oracle.com/javase/7/docs/api/java/sql/ResultSet.html 

E.g. 

* columnLabel - the label for the column specified with the SQL AS clause. If the SQL AS clause was not specified, then the label is the name of the column
* Column names used as input to getter methods are case insensitive. When a getter method is called with a column name and several columns have the same name, the value of the first matching column will be returned. The column name option is designed to be used when column names are used in the SQL query that generated the result set. For columns that are NOT explicitly named in the query, it is best to use column numbers. If column names are used, the programmer should take care to guarantee that they uniquely refer to the intended columns, which can be assured with the SQL AS clause.

Okay, we can follow it. But maybe at least document it super briefly? That this terminology is coming from X with some link perhaps etc

Comment (lsulak) [reply balta/src/main/scala/za/co/absa/db/balta/classes/QueryResultRow.scala:39]: (Basically, if we are using an underlying technology A that is using special terminology B for something that uses X and Y, and X is known thing - in our case column name - and we are not using Y, then it makes sense not to follow A's way and use B, just use X)

File: balta/src/main/scala/za/co/absa/db/balta/implicits/QueryResultRowImplicits.scala [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+        val paramType = param.typeSignature
+        val columnLabel = namingConvention.stringPerConvention(name)
+        getParamValue(columnLabel, paramType)
+      }
+
+    }
+
+    private def getParamValue[T: TypeTag](columnLabel: String, expectedType: Type): Any = {
+      val value = row(columnLabel)
+      if (isOptionType(expectedType)) {
```
Comment (benedeki) [inline balta/src/main/scala/za/co/absa/db/balta/implicits/QueryResultRowImplicits.scala:83]: Here, deciding how to pass in the actual value to the constructor.

File: balta/src/main/scala/za/co/absa/db/balta/classes/QueryResultRow.scala (hunk 5 above)
Comment (benedeki) [reply balta/src/main/scala/za/co/absa/db/balta/classes/QueryResultRow.scala:39]: Can you please suggest the documentation text. I am not sure I follow.

**Review by lsulak — APPROVED** [review]

**Pull request merged**